packages/ra-ui-materialui/src/field/ArrayField.tsx

```diff
@@ -88,86 +121,49 @@ const initialState = {
  *     )
  *     TagsField.defaultProps = { addLabel: true };
  */
-export class ArrayField extends Component<
-    FieldProps & InjectedFieldProps,
-    State
-> {
-    constructor(props: FieldProps & InjectedFieldProps) {
-        super(props);
-        this.state = props.record
-            ? this.getDataAndIds(props.record, props.source, props.fieldKey)
-            : initialState;
-    }
-
-    componentWillReceiveProps(
-        nextProps: FieldProps & InjectedFieldProps,
-        prevProps: FieldProps & InjectedFieldProps
-    ) {
-        if (nextProps.record !== prevProps.record) {
-            this.setState(
-                this.getDataAndIds(
-                    nextProps.record,
-                    nextProps.source,
-                    nextProps.fieldKey
-                )
-            );
-        }
-    }
-
-    getDataAndIds(record: object, source: string, fieldKey: string) {
-        const list = get(record, source);
-        if (!list) {
-            return initialState;
-        }
-        return fieldKey
-            ? {
-                  data: list.reduce((prev, item) => {
-                      prev[item[fieldKey]] = item;
-                      return prev;
-                  }, {}),
-                  ids: list.map(item => item[fieldKey]),
-              }
-            : {
-                  data: list.reduce((prev, item) => {
-                      prev[JSON.stringify(item)] = item;
-                      return prev;
-                  }, {}),
-                  ids: list.map(JSON.stringify),
-              };
-    }
+export const ArrayField: FunctionComponent<
+    ArrayFieldProps & FieldProps & InjectedFieldProps & State
+> = ({
+    addLabel,
+    basePath,
+    children,
+    record,
+    sortable,
+    source,
+    fieldKey,
+    ...rest
+}) => {
+    const [ids, setIds] = useState();
+    const [data, setData] = useState();
 
-    render() {
-        const {
-            addLabel,
-            basePath,
-            children,
-            record,
-            sortable,
-            source,
-            fieldKey,
-            ...rest
-        } = this.props;
-        const { ids, data } = this.state;
+    useEffect(() => {
+        const { ids, data } = getDataAndIds(record, source, fieldKey);
+        setIds(ids);
+        setData(data);
+    }, [record, source, fieldKey]);
 
-        // @ts-ignore
-        return cloneElement(Children.only(children), {
-            ids,
-            data,
-            loading: false,
-            basePath,
-            currentSort: {},
-            ...rest,
-        });
-    }
-}
+    // @ts-ignore
+    return cloneElement(Children.only(children), {
+        ids,
+        data,
+        loading: false,
+        basePath,
+        currentSort: {},
+        ...rest,
+    });
+};
 
-const EnhancedArrayField = pure<FieldProps>(ArrayField);
+const EnhancedArrayField = pure<ArrayFieldProps & FieldProps>(ArrayField);
```

Comment:
While you are updating this component to modern standards, would you mind using `React.memo` rather than `pure`?

Reply:
I tried, and I get ts errors, I revert it back to avoid to many problematic changes. But I'll try again.

Reply:
Done, but I don't know if is the best way to do it. Correct me if I'm wrong.
`memo<ArrayFieldProps & FieldProps>(ArrayFIeld);` throw me every posible TS error on `EnhancedArrayField` component.

Reply:
Done. I hope.